internal/services/compute/orchestrated_virtual_machine_scale_set_resource.go

```diff
@@ -1555,44 +1711,130 @@ func resourceOrchestratedVirtualMachineScaleSetDelete(d *pluginsdk.ResourceData,
 	return nil
 }
 
-func expandOrchestratedVirtualMachineScaleSetSkuProfile(input []interface{}) *virtualmachinescalesets.SkuProfile {
+func expandOrchestratedVirtualMachineScaleSetSkuProfile(d *pluginsdk.ResourceData, input []interface{}) *virtualmachinescalesets.SkuProfile {
 	if len(input) == 0 || input[0] == nil {
 		return nil
 	}
 
 	v := input[0].(map[string]interface{})
-	vmSizesRaw := v["vm_sizes"].(*pluginsdk.Set).List()
+	allocationStrategy := v["allocation_strategy"].(string)
 	vmSizes := make([]virtualmachinescalesets.SkuProfileVMSize, 0)
-	for _, vmSize := range vmSizesRaw {
-		vmSizes = append(vmSizes, virtualmachinescalesets.SkuProfileVMSize{
-			Name: pointer.To(vmSize.(string)),
-		})
+
+	if !features.FivePointOh() {
+		// In v4.x, we must use GetRawConfig to get only user-configured values, not merged with O+C state
+		rawConfig := d.GetRawConfig()
+		if !rawConfig.IsNull() {
+			if skuProfile := rawConfig.AsValueMap()["sku_profile"]; !skuProfile.IsNull() {
+				if skuProfiles := skuProfile.AsValueSlice(); len(skuProfiles) > 0 && !skuProfiles[0].IsNull() {
+					skuProfileConfig := skuProfiles[0].AsValueMap()
+
+					if vmSizesField := skuProfileConfig["vm_sizes"]; !vmSizesField.IsNull() {
+						vmSizesSlice := vmSizesField.AsValueSet().Values()
+						for _, vmSizeVal := range vmSizesSlice {
+							vmSizes = append(vmSizes, virtualmachinescalesets.SkuProfileVMSize{
+								Name: pointer.To(vmSizeVal.AsString()),
+							})
+						}
+					}
+
+					if vmSizeField := skuProfileConfig["vm_size"]; !vmSizeField.IsNull() {
```

Comment:
it looks like we are also setting vm_size in the code below? should we remove this and just remove the `if features.FivePointOh() {` on line 1765 instead?

Reply:
I don't believe that would work because in v4.x provider the fields are O+C, which is why I am using `d.GetRawConfig()` in the `!features.FivePointOh()` if statement, and in the `features.FivePointOh()` if statement I am using the standard `d.Get["vm_size"]` instead of the `d.GetRawConfig()`, I am only using the `d.GetRawConfig()` here to get the `rank` field because the range is from `0` to `2` and I need to verify if the value is the default Go zero value or if the `rank` field was actually set by the user or not.